### PR TITLE
[NFC][IndirectCallProm] Refactor function-based conditional devirtualization and indirect call value profile update into one helper function

### PR DESCRIPTION
**Pull request by mingmingl-llvm**
* The motivation is to move indirect callee profile update inside the function-based speculative indirect-call promotion, so that there are fewer diffs the vtable-based transformation and profile update is implemented in a follow-up patch.
* The Parent patch is https://github.com/llvm/llvm-project/pull/79381

### CODE REVIEW

**Comment by bwendling** [comment]
Please add a title and description.

**Comment by minglotus-6** [comment]
> Please add a title and description.

Done. I used a (temporary) short (and undescriptive) title to work around a lengthy branch name in spr-managed branches.  More details on why this happens is elaborated in this discourse [thread](https://discourse.llvm.org/t/do-not-use-lengthy-github-branch-name-with-getcord-spr-if-possible/76511).

I'll keep it a draft for now and send this out together with a subsequent pull request which looks cleaner with this NFC one.

File: llvm/lib/ProfileData/InstrProf.cpp [hunk 1]
```diff
@@ -1260,6 +1260,8 @@ void annotateValueSite(Module &M, Instruction &Inst,
                        ArrayRef<InstrProfValueData> VDs,
                        uint64_t Sum, InstrProfValueKind ValueKind,
                        uint32_t MaxMDCount) {
+  if (VDs.empty())
```
Comment (teresajohnson) [inline llvm/lib/ProfileData/InstrProf.cpp:1263]: Why this change?

File: llvm/lib/Transforms/Instrumentation/IndirectCallPromotion.cpp [hunk 2]
```diff
@@ -287,7 +290,18 @@ uint32_t IndirectCallPromoter::tryToPromote(
     NumOfPGOICallPromotion++;
     NumPromoted++;
   }
-  return NumPromoted;
+
+  const bool Changed = (NumPromoted != 0);
```
Comment (teresajohnson) [inline llvm/lib/Transforms/Instrumentation/IndirectCallPromotion.cpp:294]: Just return false early if NumPromoted is 0 (then Changed can be removed and just return true at the end further below).

File: llvm/lib/ProfileData/InstrProf.cpp (hunk 1 above)
Comment (teresajohnson) [reply llvm/lib/ProfileData/InstrProf.cpp:1263]: Oh, is this because you no longer check the "NumPromoted == NumVals" case?

**Review by teresajohnson — APPROVED** [review]

**Comment by minglotus-6** [comment]
Ran 'git merge main' and resolved the merge conflicts
